tools/pdf.1
.\"	$Id: pdf.1,v 1.14 2006/01/25 18:51:51 garbled Exp $
.\"
.\" Copyright (c) 1998, 1999, 2000
.\"	Tim Rightnour.  All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. All advertising materials mentioning features or use of this software
.\"    must display the following acknowledgment:
.\"	This product includes software developed by Tim Rightnour for
.\"	The NetBSD Foundation, Inc.
.\" 4. The name of Tim Rightnour may not be used to endorse or promote 
.\"    products derived from this software without specific prior written 
.\"    permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY TIM RIGHTNOUR ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL TIM RIGHTNOUR BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\" The following requests are required for all man pages.
.Dd May 5, 1999
.Dt PDF 1
.Sh NAME
.Nm pdf
.Nd display free disk space across a group of machines
.Sh SYNOPSIS
.Nm
.Op Fl ln
.Op Fl g Ar nodegroup1,...,nodegroupN
.Op Fl m Ar size
.Op Fl t Ar type
.Op Fl w Ar node1,...,nodeN
.Op Fl x Ar node1,...,nodeN
.Op Ar file | Ar file_system ...
.Sh DESCRIPTION
.Nm
displays statistics about the amount of free disk space on the specified
.Ar file_system
or on the file system of which
.Ar file
is a part on all machines in a cluster.
If neither a file or a
.Ar file_system
operand is specified,
statistics for all mounted file systems, on all machines are displayed
(subject to the
.Fl w ,
.Fl x ,
.Fl g ,
.Fl m ,
.Fl l 
and
.Fl t 
options below).
The following options are available:
.Bl -tag -width www
.It Fl l
Display statistics only about mounted file systems with the MNT_LOCAL
flag set.  If a non-local file system is given as an argument, a
warning is issued and no information is given on that file system.
.It Fl n
Print out the previously obtained statistics from the file systems.
This option should be used if it is possible that one or more
file systems are in a state such that they will not be able to provide
statistics without a long delay.
When this option is specified,
.Nm
will not request new statistics from the file systems, but will respond
with the possibly stale statistics that were previously obtained.
.It Fl g
If the
.Fl g
option is specified, followed by a comma separated list of group
names, the command will only be run on that group of nodes.  A node
may be a part of more than one group if desired, however running
without the
.Fl g
option will run the command on the same node as many times as it
appears in the file specified by the
.Ev CLUSTER
environment variable.  This option is silently ignored if used with the
.Fl w
option.
.It Fl m Ar size
Is used to limit the displayed file systems to only those with a
capacity greater than the user-supplied
.Ar size
argument.  This can be used to find filesystems over a given threshold.
.It Fl t Ar type
Is used to indicate the actions should only be taken on
filesystems of the specified type.
More than one type may be specified in a comma separated list.
The list of filesystem types can be prefixed with
.Dq no
to specify the filesystem types for which action should
.Em not
be taken.  If a file system is given on the command line that is not of
the specified type, a warning is issued and no information is given on
that file system.
.It Fl w
If the
.Fl w
option is specified, followed by a comma delimited list of machine names,
the
.Ar command
will be run on each node in the list.  Without this flag,
.Nm
runs on the nodes listed in the file pointed to by the
.Ev CLUSTER
environment variable.
.It Fl x
The
.Fl x
option can be used to exclude specific nodes from the cluster.  The
format is the same as the
.Fl w
option, a comma delimited list of machine names.  This option is
silently ignored if used with the
.Fl w
option.
.El
.Pp
.Sh ENVIRONMENT
.Nm
utilizes the following environment variables.
.Bl -tag -width "RCMD_CMD"
.It Ev CLUSTER
Contains a filename, which is a newline separated list of nodes
in the cluster.
.It Ev RCMD_CMD
Command to use to connect to remote machines.  The command chosen must
be able to connect with no password to the remote host.  Defaults to
.Ic rsh
.\" .Sh FILES
.Sh EXAMPLES
To find all the filesystems which are more than 90% full across your cluster,
you would issue:
.Bd -literal -offset indent
pdf -m 90
.Ed
.Sh DIAGNOSTICS
Exit status is 0 on success, 1 if an error occurs.
.Sh SEE ALSO
.Xr dsh 1 ,
.Xr df 1 ,
.Xr rsh 1 ,
.Xr kerberos 3 ,
.Xr hosts.equiv 5 ,
.Xr rhosts 5
.Sh HISTORY
The
.Nm
command appeared in clusterit 1.0. It is based on the
.Nm
command in IBM PSSP.
.Sh AUTHOR
.Nm Pdf
was written by Tim Rightnour.
.Sh BUGS
Output is formatted for 80 columns.  This means that mount points and
filesystems with long names 
.Po
\*(Gt 20 characters
.Pc
will be truncated.  In
addition, numbers longer than 9 digits will be truncated as well.
.Pp
.Nm
will only work on nodes that produce the
.Xr df 1
output that it expects.  It has been made to work on HP-UX, Solaris, and
AIX, along with any BSD, and most versions of Linux.

CHANGES
$Id: CHANGES,v 1.17 2006/02/01 17:52:50 garbled Exp $

* Release 2.4 *

Fix jsd to work properly on Linux.

Fix stdin command mode on all the utilities.

Make jsh and jsd more robust.  Keep jsd from permanently losing it's nodes.

Fix pdf to work on Linux.

Add a -v option to all programs to tell the user what version of
ClusterIt they are running.  CVS will allways have _BETA tagged on the end,
and it will be removed at release.

Add RVT_CMD_ARGS to dvt, to allow passing of args to rvt

Incorporate patches from debian xvt package into rvt:    
Fix xvt manpage typos
Add -rv (reverse-video) option 
Better colormap handling
All other patches were either irrelevant (makefile fixes, etc) or allready
incorporated.

Fix pdf to work on linux hosts

Replace most instances of sprintf with snprintf. Did not touch rvt, as I would
prefer to keep that as unmodified as possible, and most of them would not
be easy to fix anyhow.

Replace most instances of malloc with calloc.   Mostly to prevent any
unanticipated wierdness, and a few of the mallocs were immediately followed
by memset commands anyhow.

Add RCMD_CMD_ARGS to most of the programs. This makes it easier to do
things like run "ssh -4" as your RCMD_CMD.

Stop allocating the command to be run on the remote end as buf[MAXBUF],
instead compute it at runtime and dynamically allocate it to prevent
possible overflow on a huge command.

Fix a bug noted by Jot Powers, where trailing whitespace on a GROUP or LUMP
entry would cause dsh -g to not match it.

Fix a bug noted by John Bales  where linux machines often recieved truncated 
output from the child ssh/rsh process.

Fix bug where all the programs would mangle argv[0] for ps.

* Release 2.3.1 *

Fix a bug noted by James F Wilkus on OpenBSD where dsh would segv when
reading the clusterfile.

Convert seq over to rseq fully, not just for linux anymore.  Sigh.

Fix the clusterit.spec file to work properly.  Tested on redhat

* Release 2.3 *

Made all remote commands use poll instead of a while loop to read the incoming
stream of data from stdout and stderr.  This fixes bug where linux ssh will
hang up.

Fixed linux command line argument processing.

Changed the -l <username> processing to invoke RCMD_CMD with user@host rather
than -l user host.  This makes it much more reliable.

Added an option to dsh to test remote ports before connecting.  This can be
useful where a remote host might be down, and a long timeout would occur
waiting for a response.

Updated rvt to use a newer set of sources from xvt.  Fixes questionable
licensing in xvt, and makes it more reliable.

Converted makefiles to autoconf for ease of porting.  No more stupid patch
sets which were a terrible burden to maintain and generate.

ANSIfyed most of the code (other than rvt). Nobody uses KNF anymore.. sigh..

Incorporated many patches sent by users from the internet.  I have
unfortunately lost track of what patch came from who, but thier help is
greatly appreciated.  Most of these patches either fixed bugs, added stability
or fixed porting bugs.  The xvt conversion was based mainly on a patch sent
in by a user.  My thanks go out to those of you who have sent them in.

# changes since 2.2

* Release 2.2 *
Missing this data, sorry
* Release 2.0 *

Added dvt and rvt..  an interactive form of dsh with xterms.

Fixed a bug in jsd and dsh where it would lock up running ssh waiting for
input.

Add a jsd dir, and move jsd.c from dsh to the jsd dir.  Add nearly complete
jsh program, and jcommon.[ch].

Add new common subdir and move a tonnage of crap in there.  Mainly
common.c, common.h and strsep.c.

Fix horrid bug in dsh where dsh | more would lock terminals up.  Still
need to fix in seq, run, probably pcp.

Add <netinet/if_ether.h> to barrierd.c to shut AIX C up.

Lots of minor bugfixes, mostly dealing with failures to open files, opening
0 byte clusters, etc etc..

Pcp was still using it's own code to parse strings and groups.  Fully rewrote
the back-end internals to use the new linked-list stuff, and utilize common.c.
Pcp's code is now virtually identical to dsh, whereas before it had diverged
significantly.

Error handling has finally graduated to where I no longer feel that it
"needs work".

Fix a place where malloced data was strcatted to without being bzeroed.

Group, exclusion and -w nodes are now dynamicly generated arrays.  Removes
the MAX_GROUPS restrictions.

The nodelist and grouplists are no longer maintained in any way similar to
the original setup.  Nodes are not stored in a flat array of nodes, with a
corresponding flat array of group names.  Nodes are now stores in a linked
list of machines to run on.  Nodes which are not to be run on, are not
included in this list, so the back-end routines only need to execute code
on every node they are handed in the linked list.  This is all handled in
common.c now.  This removes the all compile time restrictions on MAXNODES,
allowing the program to scale much more efficiently.

# changes since 1.3
* Release 1.3 *

Break dsh, run and seq's common code into a common.c and common.h file,
so I don't forget to add options to run and seq when fixing dsh.  Also, fix
some locations where run and seq were out of sync slightly.

Fix a bug in pdf where if the -m option was used, filesystems at 100%
would not be displayed.

Handle ^C's correctly.  If a remote process is hung, and it is the current
process being read, you may ^C it, and the interrupt will be sent to the
remote process.  This would be just like hitting ^C on the remote end, 
and will only affect the process you are currently reading from.

Added ability to pcp multiple files and wildcards to remote hosts.

Made pcp and dsh compile on an ANSI compiler (IBM C, SUN C)

Make output from dsh look nicer, and format well.

# changes since 1.2
* Release 1.2 *

Bug fixes in dsh to deal with the start of a fangroup, and
a race condition, where large output would cause it to hang
waiting for descriptors to die.


tools/pdf.sh
#!/bin/sh
# $Id: pdf.sh,v 1.8 2006/01/25 18:51:51 garbled Exp $
args=`getopt lng:m:t:w:x: $*`
if test $? != 0
then
	echo 'Usage: pdf [-ln] [-g nodegroup] [-m size] [-t type] [-w node1,...,nodeN] [-x node1,...,nodeN] [file | file_system ...]'
	exit 2
fi
set -- $args
for i
do
	case "$i"
	in
		-l|-n)
			flag=`echo $flag $i`; shift;;
		-t)
			flag=`echo $flag $i $2`; shift; shift;;
		-w)
			warg=$2; shift; shift;;
		-g)
			garg=$2; shift; shift;;
		-x)
			xarg=$2; shift; shift;;
		-m)
			marg=$2; shift; shift;;
		--)
			shift; break;;
	esac
done

if [ -n "$warg" ]; then 
	dshargs=`echo "-w $warg"`
elif [ -n "$garg" ]; then
	dshargs=`echo "-g $garg"`
fi
if [ -n "$xarg" ]; then
	dshargs=`echo "$dshargs -x $xarg"`
fi
if [ -z "$marg" ]; then
	marg=0;
fi

(
dsh $dshargs 'sh -c "if [ `uname` = "AIX" ]; then df -kI '$flag $*'; elif [ `uname` = "HP-UX" ]; then bdf '$flag $*'; elif [ `uname` = "Linux" ]; then df -Pk '$flag $*'; else df -k '$flag $*'; fi"'
)| grep -v Filesystem | sed -e 's/://' | awk '{print $1 " " $2 " " $3 " " $4 " " $5 " " $6 " " $7}' |(
echo 'Node      Filesystem            1K-Blks     Used    Avail  Cap Mounted On'
while read node fs blocks used avail capacity mount; do \
	capacity=`echo "$capacity" | sed -e 's/\%//'`;
	if [ $capacity -gt $marg ]; then \
		printf "%-8s: %-19.19s %9d%9d%9d %3.3s%% %-17.17s\n" $node $fs $blocks $used $avail $capacity $mount; \
	fi \
done
)
